### PR TITLE
Add support for forecast-only runs on AWS

### LINKED ISSUE #2709: Make global-workflow available on AWS -- focus on ATM forecast only first
### What new functionality do you need?

Want to make global-workflow run on AWS, start with ATM forecast only.

### What are the requirements for the new functionality?

Update all submodules to make them all compile on AWS.

### Acceptance Criteria

Run global-workflow ATM only case successfully on AWS.

### Suggest a solution (optional)

_No response_

### PR DESCRIPTION
**Pull request by weihuang-jedi**
# Description
The purpose of this PR it to merge code that allowing global-workflow run on AWS, first focus on ATM forecast only.

Resolves: #2709 

# Type of change
- New feature (adds functionality)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? YES

# How has this been tested?
- Clone and build on AWS
- ATM Forecast-only on AWS

# Checklist
- [x ] Any dependent changes have been merged and published
- [x ] My code follows the style guidelines of this project
- [x ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ x] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] I have made corresponding changes to the documentation if necessary

### CODE REVIEW

File: env/AWSPW.env [hunk 1]
```diff
@@ -14,8 +14,8 @@ fi
 
 step=$1
 
-export launcher="mpiexec.hydra"
-export mpmd_opt=""
+export launcher="srun --mpi=pmi2 -l"
+export mpmd_opt="--distribution=block:block --hint=nomultithread --cpus-per-task=1"
```
Comment (aerorahul) [inline env/AWSPW.env]: This is going to be used in `atmos_products.sh` job.  Is this the right place for it?

File: modulefiles/module_gwsetup.noaacloud.lua [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+local stack_intel_ver=os.getenv("stack_intel_ver") or "2021.3.0"
+local python_ver=os.getenv("python_ver") or "3.10.3"
+
+load(pathJoin("stack-intel", stack_intel_ver))
+load(pathJoin("python", python_ver))
+load("py-jinja2")
+load("py-pyyaml")
+load("py-numpy")
+local git_ver=os.getenv("git_ver") or "1.8.3.1"
+load(pathJoin("git", git_ver))
```
Comment (aerorahul) [inline modulefiles/module_gwsetup.noaacloud.lua:18]: Following #2712, 1.8.3.1 is too old.

File: parm/config/gfs/config.resources [hunk 3]
```diff
@@ -812,7 +812,7 @@ case ${step} in
     ;;
 
   "atmos_products")
-    export wtime_atmos_products="00:15:00"
+    export wtime_atmos_products="00:45:00"
```
Comment (aerorahul) [inline parm/config/gfs/config.resources]: Going from 15 min to 45 min is a 3x increase.  Why?

File: sorc/link_workflow.sh [hunk 4]
```diff
@@ -76,6 +76,7 @@ case "${machine}" in
   "jet")      FIX_DIR="/lfs4/HFIP/hfv3gfs/glopara/git/fv3gfs/fix" ;;
   "s4")       FIX_DIR="/data/prod/glopara/fix" ;;
   "gaea")     FIX_DIR="/gpfs/f5/epic/proj-shared/global/glopara/data/fix" ;;
+  "noaacloud") FIX_DIR="/contrib/Wei.Huang/data/hack-orion/fix" ;;
```
Comment (aerorahul) [inline sorc/link_workflow.sh]: Why is this `hack-orion`?
What sort of hacks are in place?  
These should be simply synced from one of the other machines.

File: versions/run.noaacloud.ver [hunk 5]
```diff
@@ -0,0 +1,11 @@
+export stack_intel_ver=2021.3.0
+export stack_impi_ver=2021.3.0
+export spack_env=gsi-addon-env
+
+export gempak_ver=7.4.2
+
+source "${HOMEgfs:-}/versions/run.spack.ver"
+export spack_mod_path="/contrib/spack-stack/spack-stack-${spack_stack_ver}/envs/gsi-addon-env/install/modulefiles/Core"
+
+export ncl_ver=6.6.2
```
Comment (aerorahul) [inline versions/run.noaacloud.ver]: I don't think we need `ncl_ver`.

File: workflow/hosts/awspw.yaml [hunk 6]
```diff
@@ -3,11 +3,11 @@ DMPDIR: '/scratch1/NCEPDEV/global/glopara/dump' # TODO: This does not yet exist.
 PACKAGEROOT: '/scratch1/NCEPDEV/global/glopara/nwpara' #TODO: This does not yet exist.
 COMINsyn: '/scratch1/NCEPDEV/global/glopara/com/gfs/prod/syndat' #TODO: This does not yet exist.
```
Comment (aerorahul) [inline workflow/hosts/awspw.yaml]: `/scratch` filesystems are not available on AWS.  Turn them into blank and keep the `TODO`

File: workflow/hosts/awspw.yaml [hunk 7]
```diff
@@ -16,8 +16,9 @@ PARTITION_SERVICE: compute
 RESERVATION: ''
 CHGRP_RSTPROD: 'YES'
 CHGRP_CMD: 'chgrp rstprod' # TODO: This is not yet supported.
-HPSSARCH: 'YES'
+HPSSARCH: 'NO'
 HPSS_PROJECT: emc-global #TODO: See `ATARDIR` below.
+BASE_CPLIC: '/contrib/Wei.Huang/data/ICDIRS/prototype_ICs'
 LOCALARCH: 'NO'
 ATARDIR: '/NCEPDEV/${HPSS_PROJECT}/1year/${USER}/${machine}/scratch/${PSLOT}' # TODO: This will not yet work from AWS.
```
Comment (aerorahul) [inline workflow/hosts/awspw.yaml]: Turn this to blank and keep the `TODO`

File: workflow/rocoto/gefs_tasks.py [hunk 8]
```diff
@@ -11,6 +11,10 @@ def __init__(self, app_config: AppConfig, cdump: str) -> None:
 
     def stage_ic(self):
         cpl_ic = self._configs['stage_ic']
+       #The if block below is added for AWS.
+       #If we have a proper way to define 'BASE_CPLIC', this if block can be removed.
+        if ('BASE_CPLIC' not in cpl_ic.keys()):
+            cpl_ic['BASE_CPLIC'] = os.environ.get('BASE_CPLIC', '/contrib/Wei.Huang/data/ICDIRS/prototype_ICs')
```
Comment (aerorahul) [inline workflow/rocoto/gefs_tasks.py]: None of these changes should be here.
Lets discuss offline.

File: workflow/rocoto/gfs_tasks.py [hunk 9]
```diff
@@ -24,6 +24,10 @@ def stage_ic(self):
 
         # Atm ICs
         if self.app_config.do_atm:
+           #The if block below is added for AWS.
+           #If we have a proper way to define 'BASE_CPLIC', this if block can be removed.
+            if ('BASE_CPLIC' not in cpl_ic.keys()):
+                cpl_ic['BASE_CPLIC'] = os.environ.get('BASE_CPLIC', '/contrib/Wei.Huang/data/ICDIRS/prototype_ICs')
```
Comment (aerorahul) [inline workflow/rocoto/gfs_tasks.py]: ditto

File: workflow/rocoto/tasks.py [hunk 10]
```diff
@@ -214,7 +215,15 @@ def get_resource(self, task_name):
...
-            native = '--export=NONE'
+           #The PW_CSP is a AWS (CSPs parameter), if it is on CSPs, we need 'native' defined
+           #as below. Or, it won't run, but with an error:
+           #"ufs_model.x: error while loading shared libraries: libiomp5.so: cannot open shared object file: No such file or directory"
+           #Even the library path is clearly in LD_LIBRARY_PATH, or load exactly the modules when build ufs_model.x
+            pw_csp = os.environ.get('PW_CSP', 'unknown')
+            if ( pw_csp in ['aws', 'azure', 'google'] ):
+                native = '--export=ALL --exclusive'
+            else:
+                native = '--export=NONE'
```
Comment (aerorahul) [inline workflow/rocoto/tasks.py]: Please see #2701 and we can find a better way than relying on `pw_csp`.  It seems we need a mechanism to provide `native` scheduler information.

File: ush/forecast_postdet.sh [hunk 11]
```diff
@@ -268,20 +268,24 @@
     fi
   fi
 
-  # Get list of FV3 restart files
-  local file_list fv3_file
-  file_list=$(FV3_restarts)
+  ### Check that there are restart files to copy
+ #if [[ ${#restart_dates} -gt 0 ]]; then
+  if [[ -n ${restart_dates} ]]; then
```
[the remark this answers is not in the document]
Comment (aerorahul) [reply ush/forecast_postdet.sh]: This should be unnecessary if the minimum bash requirements are met.

File: workflow/rocoto/workflow_xml.py [hunk 12]
```diff
@@ -8,6 +8,7 @@
 from applications.applications import AppConfig
 from rocoto.workflow_tasks import get_wf_tasks
 import rocoto.rocoto as rocoto
+import numpy as np
```
Comment (aerorahul) [inline workflow/rocoto/workflow_xml.py]: Why `np`?
`strings` is a `list`.  `strings.append` will get you the result.

File: workflow/rocoto/workflow_xml.py [hunk 13]
```diff
@@ -156,11 +157,22 @@ def _write_crontab(self, crontab_file: str = None, cronint: int = 5) -> None:
...
-                   f'MAILTO="{replyto}"',
-                   f'{cronintstr} {rocotorunstr}',
-                   '#################################################################',
-                   '']
+                  f'#################### {pslot} ####################',
+                  f'MAILTO="{replyto}"'
+                  ]
+       #AWS need 'SHELL', and 'BASH_ENV' defined, or, the crontab job won't start.
+        pw_csp = os.environ.get('PW_CSP')
+        if ( pw_csp in ['aws', 'azure', 'google'] ):
```
Comment (aerorahul) [inline workflow/rocoto/workflow_xml.py]:
```suggestion
        if os.environ.get('PW_CSP', None) in ['aws', 'azure', 'google']:
```

File: workflow/rocoto/workflow_xml.py [hunk 14]
```diff
@@ -156,11 +157,22 @@ def _write_crontab(self, crontab_file: str = None, cronint: int = 5) -> None:
...
+             strings = np.append(strings,
```
Comment (aerorahul) [inline workflow/rocoto/workflow_xml.py]: use `strings.append` instead of `np.append`

**Review by aerorahul — CHANGES_REQUESTED** [review]
some comments regarding `mpmd_opt` and wall time for `atmos_products`.

File: env/AWSPW.env (hunk 1 above)
Comment (weihuang-jedi) [reply env/AWSPW.env]: So far it runs with forecast model.
as we define:
    export APRUN_UFS="${launcher} -n ${ntasks} ${mpmd_opt}"

for each step (fcst, atmos-products), we can remove it from some step if we do not need it.

File: modulefiles/module_gwsetup.noaacloud.lua (hunk 2 above)
Comment (weihuang-jedi) [reply modulefiles/module_gwsetup.noaacloud.lua:18]: That is what we have from "stack-stack":

[Wei.Huang@wei1epic2azure3slurm-44 global-workflow-cloud]$ module avail git

------------------- /contrib/spack-stack/spack-stack-1.6.0/envs/unified-env/install/modulefiles/intel/2021.3.0 --------------------
   git-lfs/2.4.1    git/1.8.3.1 (L)    py-gitdb/4.0.9    py-gitpython/3.1.27

If spack-stack update it, we will surely update it here.

File: parm/config/gfs/config.resources (hunk 3 above)
Comment (weihuang-jedi) [reply parm/config/gfs/config.resources]: The reason is that on Azure, and GCP, which take much longer time, on AWS, it varies, but some time exceeds 15min.
Especially, if run at /contrib (on /lustre, it is faster).

File: ush/forecast_postdet.sh (hunk 11 above)
[the remark this answers is not in the document]
Comment (weihuang-jedi) [reply ush/forecast_postdet.sh]: trying talk to NOAA-cloud, but do not expect a quick change.

File: sorc/link_workflow.sh (hunk 4 above)
Comment (weihuang-jedi) [reply sorc/link_workflow.sh]: This is copied from orion, and kept the name.
A good name/place to hold this data is welcome.

File: versions/run.noaacloud.ver (hunk 5 above)
Comment (weihuang-jedi) [reply versions/run.noaacloud.ver]: Tried remove it, will cause compare time error.
I used to be NCL developer, and do not think we need NCL for forecast or atmos-product,
but probably some cmake requirement people put in.

File: workflow/rocoto/tasks.py (hunk 10 above)
Comment (weihuang-jedi) [reply workflow/rocoto/tasks.py]: #2701 used machine name "gaea", which is in "config.resource", but here we are in python scripts, yes I need help to find a better way.

File: workflow/rocoto/workflow_xml.py (hunk 12 above)
Comment (weihuang-jedi) [reply workflow/rocoto/workflow_xml.py]: We test to remove "np".

File: parm/config/gfs/config.resources (hunk 3 above)
Comment (weihuang-jedi) [reply parm/config/gfs/config.resources]: The current code on AWS per atmos-products spend < 350s on /lustre.
(On GCP, < 1000s on /contrib). So, will reverse back on AWS.

File: workflow/rocoto/workflow_xml.py (hunk 12 above)
Comment (weihuang-jedi) [reply workflow/rocoto/workflow_xml.py]: Remove "np", and use straings.extend works.

File: workflow/rocoto/workflow_xml.py (hunk 14 above)
Comment (weihuang-jedi) [reply workflow/rocoto/workflow_xml.py]: removed "np", and using strings.extend worked.

File: sorc/link_workflow.sh (hunk 4 above)
Comment (aerorahul) [reply sorc/link_workflow.sh]: `/contrib/Wei.Huang/global/data/fix`

Comment (weihuang-jedi) [reply sorc/link_workflow.sh]: Will:

  "noaacloud") FIX_DIR="/contrib/global-workflow-shared-data/fix" ;;

better?

**Comment by weihuang-jedi** [comment]
@aerorahul I have replied your comments, and make most of the changes you requested.
There are few left:
1. remove NCL, since which cause compile error, and others have NCL as well, can we leave it as is?
2. git version, that goes with spack-stack, can we just follow spack-stack?
3. the "mpmd_opt" did not make negative impacts so far for forecast, do you want remove it from forecast and only keep in atmos-products?
4.  in tasks.py, and workflow_xml.py we use "pw_csp" local variable, some used twice and more. do you prefer not use local variable, but use "os.get_environ_get" directly.

Thanks,

Wei

File: workflow/rocoto/workflow_xml.py (hunk 13 above)
Comment (weihuang-jedi) [reply workflow/rocoto/workflow_xml.py]: changed as suggested.

File: versions/run.noaacloud.ver (hunk 5 above)
Comment (aerorahul) [reply versions/run.noaacloud.ver]: Where did you encounter compare time error?

File: env/AWSPW.env (hunk 1 above)
Comment (DavidHuber-NOAA) [reply env/AWSPW.env]: I don't think that you would want MPMD options to be sent to the UFS since the forecast job does not have an MPMD task in it.  If this change is only needed for the UFS, then I would suggest adding `--distribution=block:block --hint=nomultithread --cpus-per-task=1` to `APRUN_UFS` directly rather than assigning it globally here.

File: modulefiles/module_gwsetup.noaacloud.lua (hunk 2 above)
Comment (DavidHuber-NOAA) [reply modulefiles/module_gwsetup.noaacloud.lua:18]: Spack-stack does not install `git` but does create a module file for it based on the `git` module loaded (or system `git`) at the time of concretization.  If the system `git` is updated, then spack-stack can be reinstalled (or perhaps just reconcretized followed by `spack module lmod refresh`).

Comment (weihuang-jedi) [reply modulefiles/module_gwsetup.noaacloud.lua:18]: @DavidHuber-NOAA yes, you are right, spack-stack did not install git on AWS, but point to the system one. I'll contact NOAA-cloud and see if they can update it to a newer version.

File: env/AWSPW.env (hunk 1 above)
Comment (weihuang-jedi) [reply env/AWSPW.env]: switch mpmd_opt to "" as suggested.

File: parm/config/gfs/config.base [hunk 15]
```diff
@@ -478,4 +478,11 @@ export OFFSET_START_HOUR=0
 # Number of regional collectives to create soundings for
 export NUM_SND_COLLECTIVES=${NUM_SND_COLLECTIVES:-9}
 
+# The tracker, genesis, and METplus jobs are not supported on AWS yet
+if [[ "${machine}" == "AWSPW" ]]; then
+   export DO_TRACKER="NO"
+   export DO_GENESIS="NO"
+   export DO_METP="NO"
+fi
```
Comment (aerorahul) [inline parm/config/gfs/config.base]: This is fine for now, but we should add a `TODO` here to place these in `workflow/hosts/awspw.yaml`

Comment (weihuang-jedi) [reply parm/config/gfs/config.base]: "TODO" added as suggested.

File: versions/run.noaacloud.ver (hunk 5 above)
Comment (weihuang-jedi) [reply versions/run.noaacloud.ver]: Without set NCL, we got error at stage_ic:

Begin stage_ic.sh at Fri Jul  5 16:26:28 UTC 2024
+ stage_ic.sh[6]: . /contrib/Wei.Huang/src/global-workflow-cloud/ush/load_fv3gfs_modules.sh
++ load_fv3gfs_modules.sh[4]: [[ NO == \N\O ]]
++ load_fv3gfs_modules.sh[5]: echo 'Loading modules quietly...'
Loading modules quietly...
++ load_fv3gfs_modules.sh[6]: set +x
Lmod has detected the following error: The following module(s) are unknown:
"ncl/None"

Please check the spelling or version number. Also try "module spider ..."
It is also possible your cache file is out-of-date; it may help to try:
  $ module --ignore_cache load "ncl/None"

Also make sure that all modulefiles written in TCL start with the string
#%Module

Executing this command requires loading "ncl/None" which failed while
processing the following module(s):

    Module fullname        Module Filename
    ---------------        ---------------
    module_base.noaacloud  /contrib/Wei.Huang/src/global-workflow-cloud/modulefiles/module_base.noaacloud.lua


End stage_ic.sh at 16:26:40 with error code 1 (time elapsed: 00:00:12)

File: modulefiles/module_base.noaacloud.lua [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+local spack_mod_path=(os.getenv("spack_mod_path") or "None")
+prepend_path("MODULEPATH", spack_mod_path)
+
+load(pathJoin("stack-intel", (os.getenv("stack_intel_ver") or "None")))
+load(pathJoin("stack-intel-oneapi-mpi", (os.getenv("stack_impi_ver") or "None")))
+load(pathJoin("python", (os.getenv("python_ver") or "None")))
+
+--load(pathJoin("hpss", (os.getenv("hpss_ver") or "None")))
+load(pathJoin("gempak", (os.getenv("gempak_ver") or "None")))
+load(pathJoin("ncl", (os.getenv("ncl_ver") or "None")))
```
Comment (DavidHuber-NOAA) [inline modulefiles/module_base.noaacloud.lua]: @weihuang-jedi The `NCL` error you are seeing can be remedied by deleting this line:
```suggestion
```

Comment (weihuang-jedi) [reply modulefiles/module_base.noaacloud.lua]: Got NCL-version removed.

File: versions/run.noaacloud.ver (hunk 5 above)
Comment (weihuang-jedi) [reply versions/run.noaacloud.ver]: removed ncl-version.

File: modulefiles/module_gwsetup.noaacloud.lua (hunk 2 above)
Comment (weihuang-jedi) [reply modulefiles/module_gwsetup.noaacloud.lua:18]: Discussed with Rahul, let us go with git 1.8 on AWS for now.

Comment (aerorahul) [reply modulefiles/module_gwsetup.noaacloud.lua:18]: This PR is making use of CentOS images on AWS.  git 1.8 is the only available installation of git on this image.  Moving to a later version requires sudo commands that most of our users will be uncomfortable with at this time.
Using later git versions requires using newer AWS instances with Rocky 8 images. SS is not available on Rocky 8 at this time.  
Using git 2.3+ will be deferred till SS is available on Rocky 8 images.

File: versions/run.noaacloud.ver [hunk 17]
```diff
@@ -0,0 +1,10 @@
+export stack_intel_ver=2021.3.0
+export stack_impi_ver=2021.3.0
+export spack_env=gsi-addon-env
+
+export gempak_ver=7.4.2
```
Comment (aerorahul) [inline versions/run.noaacloud.ver]:
```suggestion
```
gempak is not available/supported on AWS/Cloud

File: modulefiles/module_base.noaacloud.lua [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+]])
+
+local spack_mod_path=(os.getenv("spack_mod_path") or "None")
+prepend_path("MODULEPATH", spack_mod_path)
+
+load(pathJoin("stack-intel", (os.getenv("stack_intel_ver") or "None")))
+load(pathJoin("stack-intel-oneapi-mpi", (os.getenv("stack_impi_ver") or "None")))
+load(pathJoin("python", (os.getenv("python_ver") or "None")))
+
+load(pathJoin("gempak", (os.getenv("gempak_ver") or "None")))
```
Comment (aerorahul) [inline modulefiles/module_base.noaacloud.lua]:
```suggestion
```
No gempak support for AWS

File: versions/run.noaacloud.ver (hunk 17 above)
Comment (weihuang-jedi) [reply versions/run.noaacloud.ver]: Removed from AWS/cloud.

File: workflow/rocoto/tasks.py (hunk 10 above)
Comment (weihuang-jedi) [reply workflow/rocoto/tasks.py]: Add a TODO note, which we need to revisit it.

File: workflow/hosts/awspw.yaml (hunk 7 above)
Comment (weihuang-jedi) [reply workflow/hosts/awspw.yaml]: Now, switch to: /bucket/global-workflow-shared-data/ICSDIR/prototype_ICs,

File: workflow/rocoto/gfs_tasks.py (hunk 9 above)
Comment (weihuang-jedi) [reply workflow/rocoto/gfs_tasks.py]: changed to /bucket/global-workflow-shared-data/ICSDIR/prototype_ICs as well.

File: workflow/hosts/awspw.yaml (hunk 6 above)
Comment (weihuang-jedi) [reply workflow/hosts/awspw.yaml]: done

File: workflow/rocoto/tasks.py [hunk 19]
```diff
@@ -176,6 +177,13 @@ def get_resource(self, task_name):
 
         task_config = self._configs[task_name]
 
+       #The PW_CSP is a AWS (CSPs parameter), if it is on CSPs,
+       #use "$USER" as account.
+        pw_csp = os.environ.get('PW_CSP', None)
+        if ( pw_csp in ['aws', 'azure', 'google'] ):
+            task_config['ACCOUNT'] = os.environ.get('USER')
+            task_config['ACCOUNT_SERVICE'] = os.environ.get('USER')
```
Comment (aerorahul) [inline workflow/rocoto/tasks.py]: AWSPW.yaml has these set to `USER`, so why is this needed here?

File: workflow/rocoto/tasks.py [hunk 20]
```diff
@@ -206,9 +214,15 @@ def get_resource(self, task_name):
...
-                memory += ':prepost=true'
+       #The PW_CSP is a AWS (CSPs parameter), if it is on CSPs, cannot define 'memory' here,
+       #Or the arch and cleanup will hang.
+        if ( pw_csp in ['aws', 'azure', 'google'] ):
+            memory = None
+        else:
+            memory = task_config.get(f'memory_{task_name}', None)
+            if scheduler in ['pbspro']:
+                if task_config.get('prepost', False):
+                    memory += ':prepost=true'
```
Comment (aerorahul) [inline workflow/rocoto/tasks.py]: This should not be necessary is `config.resources.AWSPW` unsets the memory variable similar to `Gaea` that @DavidBurrows-NCO added.

File: sorc/ufs_model.fd
Comment (aerorahul) [inline sorc/ufs_model.fd:1]: please check if the submodules need to be updated.

File: workflow/rocoto/tasks.py (hunk 19 above)
Comment (weihuang-jedi) [reply workflow/rocoto/tasks.py]: Removed. At first, did not figure out this. Thanks for point out.

File: workflow/rocoto/tasks.py (hunk 20 above)
Comment (weihuang-jedi) [reply workflow/rocoto/tasks.py]: Following Gaea example, added file config.resorces.AWSPW.

File: sorc/ufs_model.fd
Comment (weihuang-jedi) [reply sorc/ufs_model.fd:1]: tried resync with emc repo and bring submodule to update.

Comment (aerorahul) [inline sorc/ufs_model.fd:1]: Why is this module updated?

File: workflow/rocoto/tasks.py [hunk 21]
```diff
@@ -223,7 +223,16 @@ def get_resource(self, task_name):
...
+           #The PW_CSP is a AWS (CSPs parameter), if it is on CSPs, we need 'native' defined
+           #as below. Or, it won't run, but with an error:
+           #"ufs_model.x: error while loading shared libraries: libiomp5.so: cannot open shared object file: No such file or directory"
+           #Even the library path is clearly in LD_LIBRARY_PATH, or load exactly the modules when build ufs_model.x
+           #TODO: find a mechanism to provide native scheduler information.
+            pw_csp = os.environ.get('PW_CSP', 'unknown')
+            if ( pw_csp in ['aws', 'azure', 'google'] ):
+                native = '--export=ALL --exclusive'
+            else:
+                native = '--export=NONE'
```
Comment (aerorahul) [inline workflow/rocoto/tasks.py]: This should be unnecessary, if 
`export is_exclusive="True"` is set in `config.resources.AWSPW`

Comment (weihuang-jedi) [reply workflow/rocoto/tasks.py]: made this changes. and tested, it worked.
Thanks.

File: workflow/rocoto/tasks.py [hunk 22]
```diff
@@ -205,10 +206,9 @@ def get_resource(self, task_name):
         else:
             # Memory is not required
             memory = task_config.get(f'memory_{task_name}', None)
-
-        if scheduler in ['pbspro']:
-            if task_config.get('prepost', False):
-                memory += ':prepost=true'
+            if scheduler in ['pbspro']:
+                if task_config.get('prepost', False):
+                    memory += ':prepost=true'
```
Comment (aerorahul) [inline workflow/rocoto/tasks.py]: How is this change related to AWS?

File: workflow/rocoto/tasks.py [hunk 23]
```diff
@@ -223,7 +223,10 @@ def get_resource(self, task_name):
             else:
                 native += ':shared'
         elif scheduler in ['slurm']:
-            native = '--export=NONE'
+            if task_config.get('is_exclusive', False):
+                native = '--export=ALL --exclusive'
```
Comment (aerorahul) [inline workflow/rocoto/tasks.py]: Why is `--export=ALL` required for exclusive?
This is going to break every HPC

File: sorc/ufs_model.fd
Comment (weihuang-jedi) [reply sorc/ufs_model.fd:1]: my bad, I was trying compile something, and should move back the hash.

File: workflow/rocoto/tasks.py (hunk 22 above)
Comment (weihuang-jedi) [reply workflow/rocoto/tasks.py]: I do not think so. This should be from sync with EMC repo.
I do not remember make this change, and I do not think I have test anythin related to "pbspro".

File: workflow/rocoto/tasks.py (hunk 23 above)
Comment (weihuang-jedi) [reply workflow/rocoto/tasks.py]: Let me try remove --export=ALL.

File: workflow/rocoto/tasks.py (hunk 10 above)
Comment (weihuang-jedi) [reply workflow/rocoto/tasks.py]: Testing with adding "is_exclusive" in configure.resource.AWSPW now.

File: workflow/rocoto/tasks.py (hunk 23 above)
Comment (weihuang-jedi) [reply workflow/rocoto/tasks.py]: tested that remove "--export=ALL" from "native" when is_exclusive is set True in config.resource.AWSPW works fine.

File: workflow/rocoto/tasks.py (hunk 22 above)
Comment (weihuang-jedi) [reply workflow/rocoto/tasks.py]: copy over from EMC repo again, only change the native = None part.

File: workflow/rocoto/gefs_tasks.py (hunk 8 above)
Comment (weihuang-jedi) [reply workflow/rocoto/gefs_tasks.py]: Removed.

File: workflow/rocoto/gfs_tasks.py [hunk 24]
```diff
@@ -24,6 +24,12 @@ def stage_ic(self):
 
         # Atm ICs
         if self.app_config.do_atm:
+           #The if block below is added for AWS.
+           #If we have a proper way to define 'BASE_CPLIC', this if block can be removed.
```
Comment (aerorahul) [inline workflow/rocoto/gfs_tasks.py]: `BASE_CPLIC` is defined in `hosts/awspw.yaml` and that gets filled into `config.base`.

File: workflow/hosts/awspw.yaml [hunk 25]
```diff
@@ -16,10 +17,11 @@ RESERVATION: ''
 CLUSTERS: ''
 CHGRP_RSTPROD: 'YES'
 CHGRP_CMD: 'chgrp rstprod' # TODO: This is not yet supported.
-HPSSARCH: 'YES'
+HPSSARCH: 'NO'
 HPSS_PROJECT: emc-global #TODO: See `ATARDIR` below.
+BASE_CPLIC: '/bucket/global-workflow-shared-data/ICSDIR/prototype_ICs'
```
Comment (aerorahul) [inline workflow/hosts/awspw.yaml:21]: `BASE_CPLIC` is defined here for AWS.

File: workflow/rocoto/gefs_tasks.py [hunk 26]
```diff
@@ -3,7 +3,6 @@
 import rocoto.rocoto as rocoto
 from datetime import datetime, timedelta
 
-
```
Comment (aerorahul) [inline workflow/rocoto/gefs_tasks.py]: This needs to be restored.  `pynorms` is failing.

File: workflow/rocoto/gfs_tasks.py (hunk 24 above)
Comment (weihuang-jedi) [reply workflow/rocoto/gfs_tasks.py]: That is right, as BASE_CPLIC is defined in awspw.yaml now, we can removed this block.

File: workflow/rocoto/gefs_tasks.py (hunk 26 above)
Comment (weihuang-jedi) [reply workflow/rocoto/gefs_tasks.py]: copy over from develop and restore it.

File: workflow/hosts/awspw.yaml (hunk 25 above)
Comment (weihuang-jedi) [reply workflow/hosts/awspw.yaml:21]: Yes. removed the AWS notice block from hosts/gfs_tasks.py

**Comment by weihuang-jedi** [comment]
@aerorahul we need to update UPP hash to:

commit 88493f3437b4930f125a1224972fead924a0ebad
Author: Wei Huang <wei.huang@noaa.gov>
Date:   Fri Apr 26 21:29:51 2024 -0600

    add module for AWS (#934)

    * add module for AWS

    * comment that this is for NOAA Cloud
    
    what is a good way to do it?

**Review by aerorahul — DISMISSED** [review]
looks good.
My only comment is on the `--exclusive` option for `slurm` in `tasks.py` and how that may be impacts on HPC e.g. Hera.  @DavidHuber-NOAA how would you assess that?

**Comment by aerorahul** [comment]
@weihuang-jedi Please fix the errors indicated in the pynorms test.

File: env/AWSPW.env [hunk 27]
```diff
@@ -9,17 +9,205 @@ fi
...
+
+    nth_max=$((npe_node_max / npe_node_eupd))
+
+    export NTHREADS_ENKF=${nth_eupd:-${nth_max}}
+    [[ ${NTHREADS_ENKF} -gt ${nth_max} ]] && export NTHREADS_ENKF=${nth_max}
+    export APRUN_ENKF="${launcher} -n ${npe_enkf:-${npe_eupd}} --cpus-per-task=${NTHREADS_ENKF}"
+
+    export CFP_MP=${CFP_MP:-"YES"}
+    export USE_CFP=${USE_CFP:-"YES"}
+    export APRUNCFP="${launcher} -n \$ncmd ${mpmd_opt}"
```
Comment (aerorahul) [inline env/AWSPW.env]: Are any of these jobs actually running on AWS in this PR?
Can we please keep the PR limited to forecast-only?

Comment (weihuang-jedi) [reply env/AWSPW.env]: For this PR, I only tested ATM forecast only.
Do you want me remove all un-related sections?
Thank you,

File: sorc/link_workflow.sh [hunk 28]
```diff
@@ -75,7 +75,8 @@ case "${machine}" in
   "hercules") FIX_DIR="/work/noaa/global/kfriedma/glopara/fix" ;;
   "jet")      FIX_DIR="/lfs4/HFIP/hfv3gfs/glopara/git/fv3gfs/fix" ;;
   "s4")       FIX_DIR="/data/prod/glopara/fix" ;;
-  "gaea")     FIX_DIR="/gpfs/f5/ufs-ard/world-shared/global/glopara/data/fix" ;;
+  "gaea")     FIX_DIR="/gpfs/f5/epic/proj-shared/global/glopara/data/fix" ;;
```
Comment (aerorahul) [inline sorc/link_workflow.sh]: @weihuang-jedi Did the Gaea team approve of the change in the fix_dir location for GAEA?

Comment (weihuang-jedi) [reply sorc/link_workflow.sh]: @aerorahul I did not change that line myself. Those lines are coming up when I sync my branch with EMC repo (or because it did not sync well with EMC repo). 
I'll try to sync mine with EMC repo more often to see if such thing can be avoid.

File: env/AWSPW.env (hunk 27 above)
Comment (weihuang-jedi) [reply env/AWSPW.env]: I'll try to get a minimum pack for AWS with ATM forecast only.

File: sorc/link_workflow.sh (hunk 28 above)
Comment (weihuang-jedi) [reply sorc/link_workflow.sh]: OK, re-synced the code.

File: env/AWSPW.env (hunk 27 above)
Comment (weihuang-jedi) [reply env/AWSPW.env]: Now AWSPW only has a minimum packs.

**Review by aerorahul — DISMISSED** [review]
lgtm.

**Comment by aerorahul** [comment]
@weihuang-jedi 
Please fix the failing pynorms test.

File: workflow/rocoto/workflow_xml.py [hunk 29]
```diff
@@ -156,11 +156,21 @@ def _write_crontab(self, crontab_file: str = None, cronint: int = 5) -> None:
...
+            strings.extend(
+                [
+                f'SHELL="/bin/bash"',
+                f'BASH_ENV="/etc/bashrc"'
+                ])
+        strings.extend(
+            [
+            f'{cronintstr} {rocotorunstr}',
+            '#################################################################',
+            ''])
```
Comment (DavidHuber-NOAA) [inline workflow/rocoto/workflow_xml.py]: Fix `pycodestyle` issues
```suggestion
                   f'#################### {pslot} ####################',
                   f'MAILTO="{replyto}"'
                   ]
        # AWS need 'SHELL', and 'BASH_ENV' defined, or, the crontab job won't start.
        if os.environ.get('PW_CSP', None) in ['aws', 'azure', 'google']:
            strings.extend([
                            f'SHELL="/bin/bash"',
                            f'BASH_ENV="/etc/bashrc"'
                           ])
        strings.extend([
                        f'{cronintstr} {rocotorunstr}',
                        '#################################################################',
                        ''])
```

**Review by DavidHuber-NOAA — DISMISSED** [review]
Please fix the pynorms issues.

**Review by WalterKolczynski-NOAA — APPROVED** [review]
Conditionally approved

**Comment by weihuang-jedi** [comment]
@WalterKolczynski-NOAA Is there anything more that I need to do to get this PR merged?

File: workflow/rocoto/workflow_xml.py (hunk 29 above)
Comment (weihuang-jedi) [reply workflow/rocoto/workflow_xml.py]: The pynorms errors fixed.
